File: apple/Packages/MakeUpCore/Sources/MakeUpCore/Scan/ScanProtocol.swift
import Foundation

public struct ScanObservation: Sendable {
  public var time: Double, yaw: Double, pitch: Double, roll: Double, blink: Double, mouth: Double,
    faceSize: Double, brightness: Double, sharpness: Double
  public var faceCount: Int
  public var centered: Bool, tracking: Bool
  public init(
    time: Double, yaw: Double, pitch: Double, roll: Double, blink: Double, mouth: Double,
    faceSize: Double, brightness: Double, sharpness: Double, faceCount: Int, centered: Bool,
    tracking: Bool
  ) {
    self.time = time
    self.yaw = yaw
    self.pitch = pitch
    self.roll = roll
    self.blink = blink
    self.mouth = mouth
    self.faceSize = faceSize
    self.brightness = brightness
    self.sharpness = sharpness
    self.faceCount = faceCount
    self.centered = centered
    self.tracking = tracking
  }
}
public struct ScanProtocol: Sendable {
  public static let steps = ["front", "left", "right", "up", "down", "blink", "mouth", "finish"]
  public let sequence: [String]
  public private(set) var index = 0
  public private(set) var progress = 0.0
  private var since: Double?
  private var lastTime: Double?
  private var challenge = false
  public private(set) var skippedSteps: [String] = []
  @discardableResult public mutating func skipProfile() -> Bool {
    guard step == "leftProfile" || step == "rightProfile" else { return false }
    skippedSteps.append(step)
    index += 1
    resetHold()
    lastTime = nil
    return true
  }
  public var step: String { sequence[min(index, sequence.count - 1)] }
  public var complete: Bool { index == sequence.count }
  public init(side: String = "both") {
    sequence = side == "both" ? Self.steps : ["front", side == "right" ? "right" : "left",
      side == "right" ? "rightProfile" : "leftProfile", "up", "down", "blink", "mouth", "finish"]
  }
  public struct Action: Sendable {
    public var accepted: String?
    public var expression: String?
    public var issue: String?
  }
  private mutating func resetHold() {
    since = nil
    progress = 0
  }
  public mutating func update(_ o: ScanObservation) -> Action {
    if complete { return Action() }
    if let lastTime, o.time <= lastTime || o.time - lastTime > 500 {
      resetHold()
      challenge = false
    }
    lastTime = o.time
    let valid = [
      o.time, o.yaw, o.pitch, o.roll, o.blink, o.mouth, o.faceSize, o.brightness, o.sharpness,
    ].allSatisfy(\.isFinite)
    let issue: String? =
      !valid || !o.tracking || o.faceCount == 0
      ? "noFace"
      : o.faceCount != 1
        ? "oneFace"
        : !o.centered
          ? "center"
          : o.faceSize < 0.25
            ? "closer"
            : o.faceSize > 0.88
              ? "farther"
              : o.brightness < 0.16
                ? "dark"
                : o.brightness > 0.9
                  ? "bright" : o.sharpness < 8 ? "blur" : abs(o.roll) > 0.2 ? "level" : nil
    if let issue {
      resetHold()
      challenge = false
      return Action(issue: issue)
    }
    let front = abs(o.yaw) < 0.16 && abs(o.pitch) < 0.18
    let neutral = o.mouth < 0.18 && o.blink < 0.28
    let step = step
    if step == "blink" || step == "mouth" {
      if !front {
        resetHold()
        challenge = false
        return Action(issue: "front")
      }
      let active =
        step == "blink" ? o.blink > 0.6 && o.mouth < 0.18 : o.mouth > 0.5 && o.blink < 0.3
      if !challenge {
        if !active {
          resetHold()
          return Action()
        }
        if since == nil { since = o.time }
        if o.time - since! < 80 { return Action() }
        challenge = true
        resetHold()
        progress = 0.5
        return Action(expression: step)
      }
      if !neutral {
        since = nil
        progress = 0.5
        return Action(issue: "relax")
      }
    } else {
      let pose: Bool
      switch step {
      case "front", "finish": pose = front
      case "left": pose = o.yaw > 0.38 && o.yaw < 0.75 && abs(o.pitch) < 0.22
      case "right": pose = o.yaw < -0.38 && o.yaw > -0.75 && abs(o.pitch) < 0.22
      case "leftProfile": pose = o.yaw > 0.88 && o.yaw < 1.5 && abs(o.pitch) < 0.25
      case "rightProfile": pose = o.yaw < -0.88 && o.yaw > -1.5 && abs(o.pitch) < 0.25
      case "up": pose = o.pitch > 0.18 && o.pitch < 0.42 && abs(o.yaw) < 0.18
      default: pose = o.pitch < -0.16 && o.pitch > -0.4 && abs(o.yaw) < 0.18
      }
      if !pose || !neutral {
        resetHold()
        return Action(issue: neutral ? nil : "relax")
      }
    }
    if since == nil { since = o.time }
    let duration = step == "blink" || step == "mouth" ? 350.0 : 650.0
    progress = min(1, (o.time - since!) / duration)
    if progress < 1 { return Action() }
    index += 1
    challenge = false
    resetHold()
    return Action(accepted: step)
  }
}

File: web/src/features/face-scan/scanProtocol.ts
/** Guided acquisition only: expression challenges are not identity/liveness verification. */
export const SCAN_STEPS = [
  'front',
  'left',
  'right',
  'up',
  'down',
  'blink',
  'mouth',
  'finish',
] as const
export type ScanStep = (typeof SCAN_STEPS)[number] | 'leftProfile' | 'rightProfile'
export type CaptureSide = 'left' | 'right' | 'both'
export const captureSteps = (side: CaptureSide): readonly ScanStep[] => side === 'both' ? SCAN_STEPS : ['front', side, side === 'left' ? 'leftProfile' : 'rightProfile', 'up', 'down', 'blink', 'mouth', 'finish']
export interface Observation {
  time: number
  faceCount: number
  yaw: number
  pitch: number
  roll: number
  blink: number
  mouth: number
  faceSize: number
  centered: boolean
  brightness: number
  sharpness: number
  tracking: boolean
}
export class ScanProtocol {
  readonly steps: readonly ScanStep[]
  constructor(side: CaptureSide = 'both') { this.steps = captureSteps(side) }
  index = 0
  progress = 0
  private since: number | null = null
  private lastTime: number | null = null
  private challenge = false
  readonly skippedSteps: string[] = []
  skipProfile() {
    if (!['leftProfile', 'rightProfile'].includes(this.step)) return false
    this.skippedSteps.push(this.step)
    this.index++
    this.resetHold()
    this.lastTime = null
    return true
  }
  get step() {
    return this.steps[Math.min(this.index, this.steps.length - 1)]
  }
  get complete() {
    return this.index === this.steps.length
  }
  resetHold() {
    this.since = null
    this.progress = 0
  }
  update(o: Observation): {
    accepted?: ScanStep
    captureExpression?: 'blink' | 'mouth'
    issue?: string
  } {
    if (this.complete) return {}
    if (
      this.lastTime !== null &&
      (o.time <= this.lastTime || o.time - this.lastTime > 500)
    ) {
      this.resetHold()
      this.challenge = false
    }
    this.lastTime = o.time
    const issue = ![
      o.time,
      o.yaw,
      o.pitch,
      o.roll,
      o.blink,
      o.mouth,
      o.faceSize,
      o.brightness,
      o.sharpness,
    ].every(Number.isFinite)
      ? 'noFace'
      : !o.tracking || o.faceCount === 0
        ? 'noFace'
        : o.faceCount !== 1
          ? 'oneFace'
          : !o.centered
            ? 'center'
            : o.faceSize < 0.25
              ? 'closer'
              : o.faceSize > 0.88
                ? 'farther'
                : o.brightness < 0.16
                  ? 'dark'
                  : o.brightness > 0.9
                    ? 'bright'
                    : o.sharpness < 8
                      ? 'blur'
                      : Math.abs(o.roll) > 0.2
                        ? 'level'
                        : undefined
    if (issue) {
      this.resetHold()
      this.challenge = false
      return { issue }
    }
    const front = Math.abs(o.yaw) < 0.16 && Math.abs(o.pitch) < 0.18
    const neutral = o.mouth < 0.18 && o.blink < 0.28
    const step = this.step
    if (step === 'blink' || step === 'mouth') {
      if (!front) {
        this.resetHold()
        this.challenge = false
        return { issue: 'front' }
      }
      const active =
        step === 'blink'
          ? o.blink > 0.6 && o.mouth < 0.18
          : o.mouth > 0.5 && o.blink < 0.3
      if (!this.challenge) {
        if (!active) {
          this.resetHold()
          return { issue: 'performAction' }
        }
        this.since ??= o.time
        if (o.time - this.since < 80) return {}
        this.challenge = true
        this.resetHold()
        this.progress = 0.5
        return { captureExpression: step }
      }
      if (!neutral) {
        this.since = null
        this.progress = 0.5
        return { issue: 'relax' }
      }
    } else {
      const pose =
        step === 'front' || step === 'finish'
          ? front
          : step === 'left'
            ? o.yaw > 0.38 && o.yaw < 0.75 && Math.abs(o.pitch) < 0.22
            : step === 'right'
              ? o.yaw < -0.38 && o.yaw > -0.75 && Math.abs(o.pitch) < 0.22
              : step === 'leftProfile' ? o.yaw > 0.88 && o.yaw < 1.5 && Math.abs(o.pitch) < 0.25
              : step === 'rightProfile' ? o.yaw < -0.88 && o.yaw > -1.5 && Math.abs(o.pitch) < 0.25
              : step === 'up'
                ? o.pitch > 0.18 && o.pitch < 0.42 && Math.abs(o.yaw) < 0.18
                : o.pitch < -0.16 && o.pitch > -0.4 && Math.abs(o.yaw) < 0.18
      if (!pose || !neutral) {
        this.resetHold()
        return { issue: !neutral ? 'relax' : 'adjustPose' }
      }
    }
    this.since ??= o.time
    const duration = step === 'blink' || step === 'mouth' ? 350 : 650
    this.progress = Math.min(1, (o.time - this.since) / duration)
    if (this.progress < 1) return {}
    this.index++
    this.challenge = false
    this.resetHold()
    return { accepted: step }
  }
}
